File: in_a_Weekend/python/vec3.py
import math

class Vec3(object):

    def __init__(self, e0=0, e1=0, e2=0):
        self.e = [e0, e1, e2]

    @property
    def x(self):
        return self.e[0]

    @property
    def y(self):
        return self.e[1]

    @property
    def z(self):
        return self.e[2]

    @property
    def r(self):
        return self.e[0]

    @property
    def g(self):
        return self.e[1]

    @property
    def b(self):
        return self.e[2]

    def __pos__(self):
        return self

    def __neg__(self):
        x, y, z = self.e
        return Vec3(-x, -y, -z)

    def __getitem__(self, idx):
        return self.e[idx]

    def __setitem__(self, idx, value):
        self.e[idx] = value

    def __add__(self, v2):
        x1, y1, z1 = self.e
        x2, y2, z2 = v2.e
        return Vec3(x1+x2, y1+y2, z1+z2)

    def __sub__(self, v2):
        x1, y1, z1 = self.e
        x2, y2, z2 = v2.e
        return Vec3(x1-x2, y1-y2, z1-z2)

    def __mul__(self, v2):
        x1, y1, z1 = self.e
        if isinstance(v2, Vec3):
            x2, y2, z2 = v2.e
            return Vec3(x1*x2, y1*y2, z1*z2)
        else:
            return Vec3(x1*v2, y1*v2, z1*v2)

    def __rmul__(self, v2):
        return self.__mul__(v2)

    def __div__(self, v2):
        x1, y1, z1 = self.e
        if isinstance(v2, Vec3):
            x2, y2, z2 = v2.e
            return Vec3(x1/float(x2), y1/float(y2), z1/float(z2))
        else:
            return Vec3(x1/float(v2), y1/float(v2), z1/float(v2))

    def __eq__(self, v2):
        x1, y1, z1 = self.e
        x2, y2, z2 = v2.e
        def close(a, b):
            return abs(a-b) <= 1e-8
        return close(x1, x2) and close(y1, y2) and close(z1, z2)

    @property
    def length(self):
        return math.sqrt(self.dot(self))

    @property
    def unitVector(self):
        return self / self.length

    def dot(self, v2):
        x1, y1, z1 = self.e
        x2, y2, z2 = v2.e
        return x1*x2 + y1*y2 + z1*z2

    def cross(self, v2):
        x1, y1, z1 = self.e
        x2, y2, z2 = v2.e
        return Vec3(
            y1*z2 - z1*y2,
            -(x1*z2 - z1*x2),
            x1*y2 - y1*x2
        )

    def update(self, other):
        self.e[0] = other.e[0]
        self.e[1] = other.e[1]
        self.e[2] = other.e[2]

    @staticmethod
    def Unit(v):
        return v.unitVector

    @staticmethod
    def Dot(v1, v2):
        return v1.dot(v2)

    @staticmethod
    def Cross(v1, v2):
        return v1.cross(v2)

    def __repr__(self):
        return 'Vec3({}, {}, {})'.format(self.e[0], self.e[1], self.e[2])

if __name__ == '__main__':
    v = Vec3()
    assert v[0] == 0 and v[1] == 0 and v[2] == 0

    v = Vec3(1, 2, 3)
    assert v[0] == 1 and v[1] == 2 and v[2] == 3

    v[0], v[1], v[2] = 4, 5, 6
    assert v[0] == 4 and v[1] == 5 and v[2] == 6

    v = Vec3(1, 2, 3)
    assert v.x == 1 and v.y == 2 and v.z == 3
    assert v.r == 1 and v.g == 2 and v.b == 3

    v = +v
    assert v.x == 1 and v.y == 2 and v.z == 3

    v = -v
    assert v.x == -1 and v.y == -2 and v.z == -3

    a, b = Vec3(1, 2, 3), Vec3(4, 5, 6)
    c = a + b
    assert c == Vec3(5, 7, 9)

    c = a - b
    assert c == Vec3(-3, -3, -3)

    c = a * b
    assert c == Vec3(4, 10, 18)

    c = a / b
    assert c == Vec3(0.25, 0.4, 0.5)

    c = a * 2
    assert c == Vec3(2, 4, 6)

    c = 2 * a
    assert c == Vec3(2, 4, 6)

    c = a / 2
    assert c == Vec3(0.5, 1, 1.5)

    a += b
    assert a == Vec3(5, 7, 9)

    a -= b
    assert a == Vec3(1, 2, 3)

    a *= b
    assert a == Vec3(4, 10, 18)

    a /= b
    assert a == Vec3(1, 2, 3)

    a *= 3
    assert a == Vec3(3, 6, 9)

    a /= 3
    assert a == Vec3(1, 2, 3)

    a = Vec3(3, 0, -4)
    assert a.length == 5
    assert a.unitVector == Vec3(0.6, 0, -0.8)

    c = a.dot(b)
    assert c == -12

    c = a.cross(b)
    assert c == Vec3(20, -34, 15)
    print 'ok'

File: in_a_Weekend/rust/rti1w/src/vec3.rs
use std::f32;
use std::ops::*;

/*************************************************************************************************/
#[derive(Debug, PartialEq)]
pub struct Color { pub r: u8, pub g: u8, pub b: u8 }

impl Color {
    pub fn from_vec3(v: &Vec3) -> Self {
        fn u(f: f32) -> u8 { (255.99*f) as u8 }
        Color { r: u(v.x), g: u(v.y), b: u(v.z) }
    }
}

/*************************************************************************************************/
#[derive(Debug, Clone, Copy)]
pub struct Vec3 { pub x: f32, pub y: f32, pub z: f32 }

impl Vec3 {
    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Vec3 { x, y, z }
    }
    pub fn zero() -> Self {
        Vec3 { x: 0., y: 0., z: 0. }
    }
    pub fn dot(&self, other: Self) -> f32 { self.x*other.x + self.y*other.y + self.z*other.z }
    pub fn cross(&self, other: Self) -> Self {
        Vec3 {
            x: self.y*other.z - self.z*other.y,
            y: self.z*other.x - self.x*other.z,
            z: self.x*other.y - self.y*other.x
        }
    }
    pub fn length(&self) -> f32 { self.dot(*self).sqrt() }
    pub fn unit(v: Vec3) -> Vec3 { v / v.length() }
}

// v1 = v2
impl PartialEq for Vec3 {
    fn eq(&self, other: &Vec3) -> bool {
        (self.x-other.x).abs() <= f32::EPSILON &&
        (self.y-other.y).abs() <= f32::EPSILON &&
        (self.z-other.z).abs() <= f32::EPSILON
    }
}
impl Eq for Vec3 {}

// v1 + v2
impl Add for Vec3 {
    type Output = Vec3;
    fn add(self, other: Self) -> Vec3 {
        Vec3 {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z
        }
    }
}

// v1 += v2
impl AddAssign for Vec3 {
    fn add_assign(&mut self, other: Vec3) {
        *self = Vec3 {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z
        }
    }
}

// -v
impl Neg for Vec3 {
    type Output = Self;
    fn neg(self) -> Self{
        Vec3 {
            x: -self.x,
            y: -self.y,
            z: -self.z
        }
    }
}

// v1 - v2
impl Sub for Vec3 {
    type Output = Self;
    fn sub(self, other: Self) -> Self {
        Vec3 {
            x: self.x - other.x,
            y: self.y - other.y,
            z: self.z - other.z
        }
    }
}

// v1 -= v2
impl SubAssign for Vec3 {
    fn sub_assign(&mut self, other: Vec3) {
        *self = Vec3 {
            x: self.x - other.x,
            y: self.y - other.y,
            z: self.z - other.z
        }
    }
}

// v1 * v2
impl Mul for Vec3 {
    type Output = Self;
    fn mul(self, other: Self) -> Self {
        Vec3 {
            x: self.x * other.x,
            y: self.y * other.y,
            z: self.z * other.z
        }
    }
}

// v1 *= v2
impl MulAssign for Vec3 {
    fn mul_assign(&mut self, other: Vec3) {
        *self = Vec3 {
            x: self.x * other.x,
            y: self.y * other.y,
            z: self.z * other.z
        }
    }
}

// v * a
impl Mul<f32> for Vec3 {
    type Output = Vec3;
    fn mul(self, other: f32) -> Vec3 {
        Vec3 {
            x: self.x * other,
            y: self.y * other,
            z: self.z * other
        }
    }
}

// v *= a
impl MulAssign<f32> for Vec3 {
    fn mul_assign(&mut self, other: f32) {
        *self = Vec3 {
            x: self.x * other,
            y: self.y * other,
            z: self.z * other
        }
    }
}

// a * v
impl Mul<Vec3> for f32 {
    type Output = Vec3;
    fn mul(self, other: Vec3) -> Vec3 {
        Vec3 {
            x: self * other.x,
            y: self * other.y,
            z: self * other.z
        }
    }
}

// v1 / v2
impl Div for Vec3 {
    type Output = Self;
    fn div(self, other: Self) -> Self {
        Vec3 {
            x: self.x / other.x,
            y: self.y / other.y,
            z: self.z / other.z
        }
    }
}

// v1 /= v2
impl DivAssign for Vec3 {
    fn div_assign(&mut self, other: Vec3) {
        *self = Vec3 {
            x: self.x / other.x,
            y: self.y / other.y,
            z: self.z / other.z
        }
    }
}

// v / a
impl Div<f32> for Vec3 {
    type Output = Vec3;
    fn div(self, other: f32) -> Vec3 {
        Vec3 {
            x: self.x / other,
            y: self.y / other,
            z: self.z / other
        }
    }
}

// v /= a
impl DivAssign<f32> for Vec3 {
    fn div_assign(&mut self, other: f32) {
        *self = Vec3 {
            x: self.x / other,
            y: self.y / other,
            z: self.z / other
        }
    }
}

#[cfg(test)]
mod tests {
    use std::f32;
    use super::*;

    #[test]
    fn test_color() {
        let c = Color::from_vec3(&Vec3::zero());
        assert_eq!(c, Color { r: 0, g: 0, b: 0 });

        let c = Color::from_vec3(&Vec3::new(1.,1.,1.));
        assert_eq!(c, Color { r: 255, g: 255, b: 255 });
    }

    #[test]
    fn test_create() {
        let a = Vec3::new(1., 2., 3.);
        assert_eq!(a.x, 1.);
        assert_eq!(a.y, 2.);
        assert_eq!(a.z, 3.);
    }

    #[test]
    fn test_create_zero() {
        let a = Vec3::zero();
        assert_eq!(a.x, 0.);
        assert_eq!(a.y, 0.);
        assert_eq!(a.z, 0.);
    }

    #[test]
    fn test_length() {
        let a = Vec3::new(1., 2., 3.);
        assert!((a.length()-14_f32.sqrt()).abs() <= f32::EPSILON);
    }

    #[test]
    fn test_unit() {
        let a = Vec3::new(2., 0., 0.);
        assert_eq!(Vec3::unit(a), Vec3::new(1., 0., 0.));
    }

    #[test]
    fn test_dot() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a.dot(b), 32.);
    }

    #[test]
    fn test_cross() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a.cross(b), Vec3::new(-3., 6., -3.));
    }

    #[test]
    fn test_add() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a+b, Vec3::new(5., 7., 9.));
    }

    #[test]
    fn test_add_assign() {
        let mut a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        a += b;
        assert_eq!(a, Vec3::new(5., 7., 9.));
    }

    #[test]
    fn test_neg() {
        let a = Vec3::new(1., 2., 3.);
        assert_eq!(-a, Vec3::new(-1., -2., -3.));
    }

    #[test]
    fn test_sub() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a-b, Vec3::new(-3., -3., -3.));
    }

    #[test]
    fn test_sub_assign() {
        let mut a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        a -= b;
        assert_eq!(a, Vec3::new(-3., -3., -3.));
    }

    #[test]
    fn test_mul() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a*b, Vec3::new(4., 10., 18.));
    }

    #[test]
    fn test_mul_assign() {
        let mut a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        a *= b;
        assert_eq!(a, Vec3::new(4., 10., 18.));
    }

    #[test]
    fn test_mul_scalar() {
        let a = Vec3::new(1., 2., 3.);
        assert_eq!(a*2., Vec3::new(2., 4., 6.));
    }

    #[test]
    fn test_mul_assign_scalar() {
        let mut a = Vec3::new(1., 2., 3.);
        a *= 2.;
        assert_eq!(a, Vec3::new(2., 4., 6.));
    }

    #[test]
    fn test_mul_scalar_l() {
        let a = Vec3::new(1., 2., 3.);
        assert_eq!(2.*a, Vec3::new(2., 4., 6.));
    }

    #[test]
    fn test_div() {
        let a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        assert_eq!(a/b, Vec3::new(0.25, 0.4, 0.5));
    }

    #[test]
    fn test_div_assign() {
        let mut a = Vec3::new(1., 2., 3.);
        let b = Vec3::new(4., 5., 6.);
        a /= b;
        assert_eq!(a, Vec3::new(0.25, 0.4, 0.5));
    }

    #[test]
    fn test_div_scalar() {
        let a = Vec3::new(1., 2., 3.);
        assert_eq!(a/2., Vec3::new(0.5, 1., 1.5));
    }

    #[test]
    fn test_div_assign_scalar() {
        let mut a = Vec3::new(1., 2., 3.);
        a /= 2.;
        assert_eq!(a, Vec3::new(0.5, 1., 1.5));
    }

}
